src/sync.rs

```diff
@@ -0,0 +1,302 @@
+struct Condvar {
+    waiters: VecDeque<CondvarWaiter>,
+}
+
+/// The state of all synchronization variables.
+#[derive(Default, Debug)]
+pub(super) struct SynchronizationState {
+    mutexes: IndexVec<MutexId, Mutex>,
+    rwlocks: IndexVec<RwLockId, RwLock>,
+    condvars: IndexVec<CondvarId, Condvar>,
+}
+
+// Public interface to synchronization primitives. Please note that in most
+// cases, the function calls are infallible and it is the client's (shim
+// implementation's) responsibility to detect and deal with erroneous
+// situations.
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    #[inline]
+    /// Create state for a new mutex.
+    fn mutex_create(&mut self) -> MutexId {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes.push(Default::default())
+    }
+
+    #[inline]
+    /// Get the id of the thread that currently owns this lock.
+    fn mutex_get_owner(&mut self, id: MutexId) -> ThreadId {
+        let this = self.eval_context_ref();
+        this.machine.threads.sync.mutexes[id].owner.unwrap()
+    }
+
+    #[inline]
+    /// Check if locked.
+    fn mutex_is_locked(&mut self, id: MutexId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].owner.is_some()
+    }
+
+    /// Lock by setting the mutex owner and increasing the lock count.
+    fn mutex_lock(&mut self, id: MutexId, thread: ThreadId) {
+        let this = self.eval_context_mut();
+        let mutex = &mut this.machine.threads.sync.mutexes[id];
+        if let Some(current_owner) = mutex.owner {
+            assert_eq!(thread, current_owner, "mutex already locked by another thread");
+            assert!(
+                mutex.lock_count > 0,
+                "invariant violation: lock_count == 0 iff the thread is unlocked"
+            );
+        } else {
+            mutex.owner = Some(thread);
+        }
+        mutex.lock_count = mutex.lock_count.checked_add(1).unwrap();
+    }
+
+    /// Unlock by decreasing the lock count. If the lock count reaches 0, unset
+    /// the owner.
+    fn mutex_unlock(&mut self, id: MutexId) -> Option<(ThreadId, usize)> {
+        let this = self.eval_context_mut();
+        let mutex = &mut this.machine.threads.sync.mutexes[id];
+        if let Some(current_owner) = mutex.owner {
+            mutex.lock_count = mutex
+                .lock_count
+                .checked_sub(1)
+                .expect("invariant violation: lock_count == 0 iff the thread is unlocked");
+            if mutex.lock_count == 0 {
+                mutex.owner = None;
+            }
+            Some((current_owner, mutex.lock_count))
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_enqueue(&mut self, id: MutexId, thread: ThreadId) {
```

Comment:
The doc comment doesn't match the the function name (and the function name seems to be right).